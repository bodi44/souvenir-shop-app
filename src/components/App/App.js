import React from 'react'
import PropTypes from 'prop-types'

import './App.scss'
import BEM from '../../helpers/BEM'

const b = BEM('App')

const App = props => (
  <div className={b()}>
    <h1 className={b('title')}>Souvenir-store App</h1>
    {props.message ? <div className={b('data-error')}>{props.message}</div> : props.children}
  </div>
)

App.propTypes = {
  message: PropTypes.string,
  children: PropTypes.node
}

export default App