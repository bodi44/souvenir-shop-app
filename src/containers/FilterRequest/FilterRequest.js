import React, { useState } from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'

import {
  clearPurchasesByDate,
  purchase,
  reportAboutPurchases,
  showAllPurchases,
  showErrorMessage,
  showHelpMessage
} from '../../actions/souvenirStore'
import { getAllPurchasesByDate } from '../../reducers'

import RequestForm from '../../components/RequestForm'

import isValidDate from '../../helpers/isValidDate'

const FilterRequest = props => {
  const {
    allPurchases,
    currencies,
    purchase,
    showAllPurchases,
    clearPurchasesByDate,
    reportAboutPurchases,
    showErrorMessage,
    showHelpMessage
  } = props

  const [input, setInput] = useState('')
  const [buttonColor, setButtonColor] = useState(false)

  const checkPurchaseCommand = (date, price, currency, productName) => {
    if (!isValidDate(new Date(date)) || currencies[currency] === undefined) {
      showErrorMessage('Error: Invalid arguments given.')
      return
    }

    purchase(date, price, currency, productName)
  }

  const checkClearCommand = date => {
    if (!isValidDate(new Date(date))) {
      showErrorMessage('Error: Invalid arguments given.')
      return
    }

    clearPurchasesByDate(date)
  }

  const report = (year, currency) => {
    if (currencies[currency] === undefined || !isValidDate(new Date(year))) {
      showErrorMessage('Error: Invalid arguments given.')
      return
    }

    let totalIncome = 0
    const filteredData = allPurchases.filter(purchaseDate => purchaseDate.id.slice(0, 4) === year)

    filteredData.map(date => date.purchases.forEach(purchase => {
        purchase.currency !== currency ?
          totalIncome = totalIncome + purchase.price / currencies[purchase.currency] * currencies[currency]
          :
          totalIncome += purchase.price
      }
    ))

    reportAboutPurchases(`${Number((totalIncome).toFixed(2))} ${currency}`)
  }

  const methods = {
    'purchase': { callback: checkPurchaseCommand, argLength: 4 },
    'all': { callback: showAllPurchases, argLength: 0 },
    'clear': { callback: checkClearCommand, argLength: 1 },
    'report': { callback: report, argLength: 2 },
    'help': { callback: showHelpMessage, argLength: 0 }
  }

  const handleInputChange = e => {
    e.preventDefault()
    setInput(e.target.value)
  }

  const handleSubmit = e => {
    e.preventDefault()
    let args = input.split(' ')

    if (methods.hasOwnProperty(args[0])) {
      if (methods[args[0]].callback === checkPurchaseCommand && args.slice(1).filter(arg => arg !== '').length > 4) {
        const re = /(^"|"$)/
        let string = ''

        if (args.slice(1)[3].match(re)) {
          args.slice(4).forEach(arg => string += ` ${arg}`)
        }

        if (!string.match(re)) {
          showErrorMessage('Error: invalid number of arguments given')
          return
        }

        methods[args[0]].callback(...args.slice(1, 4), string)

      } else {
        methods[args[0]].argLength === args.slice(1).filter(arg => arg !== '').length ?
          methods[args[0]].callback(...args.slice(1)) :
          showErrorMessage('Error: invalid number of arguments given')
      }

    } else {
      showErrorMessage('Error: No such command.')
    }
  }

  const handleButtonColorChange = () => {
    setButtonColor(!buttonColor)
  }

  const buttonStyle = buttonColor ? '#7f8fa6' : '#bdc3c7'

  return (
    <RequestForm
      buttonStyle={buttonStyle}
      handleInputChange={handleInputChange}
      handleSubmit={handleSubmit}
      handleButtonColorChange={handleButtonColorChange}
    />
  )
}

FilterRequest.propTypes = {
  allPurchases: PropTypes.array,
  currencies: PropTypes.object,
  purchase: PropTypes.func,
  showAllPurchases: PropTypes.func,
  clearPurchasesByDate: PropTypes.func,
  reportAboutPurchases: PropTypes.func,
  showErrorMessage: PropTypes.func,
  showHelpMessage: PropTypes.func
}

export default connect(
  state => ({
    allPurchases: getAllPurchasesByDate(state)
  }),
  {
    purchase,
    showAllPurchases,
    clearPurchasesByDate,
    reportAboutPurchases,
    showErrorMessage,
    showHelpMessage
  }
)(FilterRequest)