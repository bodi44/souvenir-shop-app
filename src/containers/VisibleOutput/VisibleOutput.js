import React from 'react'
import { connect } from 'react-redux'
import PropTypes from 'prop-types'

import {
  getAllPurchasesByDate,
  getTotalIncome,
  isStoreHasErrorMessage,
  showHelpMessage
} from '../../reducers'
import Output from '../../components/Output/'
import HelpMessage from '../../components/HelpMessage/HelpMessage'

const VisibleOutput = ({ allPurchases, totalIncome, storeErrorMessage, helpMessage }) => {
  if (storeErrorMessage) {
    return (
      <Output title={'Error:'} error={'error'}>
        {storeErrorMessage} <br/> To get instructions please type: <strong>help</strong>
      </Output>
    )
  }

  if (totalIncome) {
    return (
      <Output title={'Output:'} output={totalIncome}/>
    )
  }

  if (helpMessage) {
    return (
      <Output title={'Help:'}>
        <HelpMessage/>
      </Output>
    )
  }

  return (
    <Output title={'Output:'} allPurchases={allPurchases}/>
  )
}

Output.propTypes = {
  allPurchases: PropTypes.array.isRequired,
  totalIncome: PropTypes.string,
  storeErrorMessage: PropTypes.string,
  helpMessage: PropTypes.bool
}

export default connect(
  state => ({
    allPurchases: getAllPurchasesByDate(state),
    totalIncome: getTotalIncome(state),
    storeErrorMessage: isStoreHasErrorMessage(state),
    helpMessage: showHelpMessage(state)
  }),
  null
)(VisibleOutput)
