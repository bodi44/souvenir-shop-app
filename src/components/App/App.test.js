import React from 'react'
import Enzyme, { shallow } from 'enzyme'
import expect from 'expect'
import Adapter from 'enzyme-adapter-react-16'

import App from '../App'
import FilterRequest from '../../containers/FilterRequest/FilterRequest'
import VisibleOutput from '../../containers/VisibleOutput/VisibleOutput'

Enzyme.configure({ adapter: new Adapter() })

describe('App component', () => {
  it('should render correctly with given message props', () => {
    const component = shallow(
      <App message={'Data is Loading...'}/>
    )
    expect(component).toMatchSnapshot()
  })

  it('should render correctly with no props but with React children', () => {
    const currencies = {
      'BTC': 0.000211,
      'CZK': 25.640122,
      'EUR': 1,
      'GBP': 0.860335,
      'UAH': 29.569446,
      'USD': 1.120097
    }

    const component = shallow(
      <App>
        <FilterRequest currencies={currencies}/>
        <VisibleOutput/>
      </App>
    )
    expect(component).toMatchSnapshot()
  })
})