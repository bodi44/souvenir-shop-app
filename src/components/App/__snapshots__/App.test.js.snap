// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App component should render correctly with given message props 1`] = `ShallowWrapper {}`;

exports[`App component should render correctly with no props but with React children 1`] = `ShallowWrapper {}`;
